source/basic.tex

```diff
@@ -4589,14 +4589,14 @@
 \grammarterm{new-type-id}\iref{expr.new} when the object is created.
 \begin{itemize}
 \item A \defnadj{const}{object} is an object of type \tcode{const T} or a
-  non-mutable subobject of such an object.
+  non-mutable subobject of a const object.
 
 \item A \defnadj{volatile}{object} is an object of type
-  \tcode{volatile T}, a subobject of such an object, or a mutable
+  \tcode{volatile T}, a subobject of a volatile object, or a mutable
   subobject of a const volatile object.
```

Comment:
The "or a mutable subobject of a const volatile object" part is redundant, isn't it? A const volatile object is a volatile object, so such an object would already qualify as "a subobject of a volatile object".

Reply:
@zygoloid, this logic seems somehow flawed: You seem to argue that any subobject of a const volatile object is a volatile object (regardless of "mutable"), simply because a const volatile object is-a volatile object. This seems obviously wrong, because it drops const-ness. I think the issue is the premise: A const volatile object is not a volatile object, it's something else.

Reply:
I think what I described is the consequence of the current wording. An object of type `const volatile int` is a const object because it's of type `const T`, where `T` is `volatile int`, and it's a volatile object because it's of type `volatile T` where `T` is `const int`.

If we want to define "const volatile object" such that a const object is not a const volatile object and a volatile object is not a const volatile object, then, in addition to fixing this wording to clarify that `T` is cv-unqualified, we need to fix most mentions of "const object" in the standard to say "const object or const volatile object" (and similarly for the handful of mentions of "volatile object"), and should replace "non-volatile const object" with "(non-volatile) const object" since the non-volatile part is implied. I'd be OK with doing that, but it's not my preference.

My preference is to make the terms match normal English usage for an adjective: that is, a const volatile object is a kind of const object and is a kind of volatile object (and all of these are kinds of object, of course). That's also how almost all the rest of the standard currently uses the terms, as far as I can see. (In fact, "const volatile object" only appears twice, and in both times it's in a list of "const object, volatile object, or const volatile object", which could be rephrased as "const and/or volatile object" or similar.)

Reply:
@zygoloid: Ok, but that means more surgery to these words, beyond the antecedent clarification requested by Hubert.